src/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerable.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Threading;
+
+namespace System.IO.Enumeration
+{
+    public class FileSystemEnumerable<TResult> : IEnumerable<TResult>
+    {
+        private DelegateEnumerator _enumerator;
+        private readonly FindTransform _transform;
+        private readonly EnumerationOptions _options;
+        private readonly string _directory;
+
+        public FileSystemEnumerable(string directory, FindTransform transform, EnumerationOptions options = null)
+        {
+            _directory = directory ?? throw new ArgumentNullException(nameof(directory));
+            _transform = transform ?? throw new ArgumentNullException(nameof(transform));
+            _options = options ?? EnumerationOptions.Default;
+
+            // We need to create the enumerator up front to ensure that we throw I/O exceptions for
+            // the root directory on creation of the enumerable.
+            _enumerator = new DelegateEnumerator(this);
+        }
+
+        public FindPredicate ShouldIncludePredicate { get; set; }
+        public FindPredicate ShouldRecursePredicate { get; set; }
+
+        public IEnumerator<TResult> GetEnumerator()
+        {
+            DelegateEnumerator enumerator = Interlocked.Exchange(ref _enumerator, null);
+            if (enumerator == null)
+                enumerator = new DelegateEnumerator(this);
+
+            return enumerator;
+        }
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        /// <summary>
+        /// Delegate for filtering out find results.
+        /// </summary>
+        public delegate bool FindPredicate(ref FileSystemEntry entry);
+
+        /// <summary>
+        /// Delegate for transforming raw find data into a result.
+        /// </summary>
+        public delegate TResult FindTransform(ref FileSystemEntry entry);
+
+        private class DelegateEnumerator : FileSystemEnumerator<TResult>
```

Comment:
We should seal types like this, in particular in case the JIT is then able to do better at devirtualization.